### DISCUSSION TITLE
Layout artifacts

### DISCUSSION THREAD

**Discussion by tomlm** [Consolonia/Consolonia#129] [General]
I don't understand why sometimes controls layout perfectly, and sometimes you get this cascading layout error.

![image](https://github.com/user-attachments/assets/21ac238e-d374-4dc1-8c7e-9a7b31ec52fb)

It's not consistent in how it manifests which makes me think it's a race condition or concurrency issue.

What thoughts do you guys have? It's driving me a bit nuts.

**Comment by tomlm** [comment]
OK, I have isolated it a bit.  It seems the artifacts occur anytime there is a ScrollViewer in play. **including a combobox**

It has nothing to do with resizing.

What I did was made ScrollViewer the default gallery view.  It comes with artifacts. I then enabled/disabled every control until artifacts went away, and they go away when there is no combobox or scrollviewer. 

I think there is something about ScrollViewer which is throwing a monkey wrench into the layout engine.

**Reply by tomlm** [reply]
if you notice, there is a left shift on the Horizontal combo, on the vertical combo, on the start of the vertical scrollbar at the end of the vertical scrollbar and 2 for the start and end of the horizontal scrollbar.

![image](https://github.com/user-attachments/assets/a780558b-deef-41a4-9448-af58bd454f2f)

**Comment by tomlm** [comment]
Another observation. Artifacts are left on the screen because only invalidated rects get redrawn. When something does overwrite them without the presence of the scrollbar it draws the correct values.

This gives the impression that layout is broken all over the place, but I think it's mostly the artifacts written to the pixelbuffer when scrollbars are present.

**Comment by tomlm** [comment]
Another data point, I commented out the ScrollBar style and I don't see any jaggies anymore. (no scrollbars either of course).

The only artifact I see is on the combobox gallery page, there are a couple of stray chars. Don't know what that's about.  At any rate, I'm pretty convinced the scrollbar is breaking things pretty bad.

**Comment by jinek** [comment]
Similar I have seen was the issue with special symbols which took more than 1 character widths. Or opposite, two characters were recognized by terminal as one.
To debug you can stop on the breakpoint when printing to the console and check which symbols are printed.

So my point it's not related to layout at all, probably it's rendering issue.

**Reply by tomlm** [reply]
ah, I can try that.

**Reply by tomlm** [reply]
I found the bug.  It's because we are attempting to track the caret position using _headPosition.  When we call SetCaretPosition, if we think we are already in the write place we don't set the caret position. When we write the last char out, we update the cached location to be the next line, then when we call SetCaretPosition it doesn't (it's still in the old position) and so the line writes on the previous line wrapping around. This error cascades.

Removing _headPosition (aka the attempt to bypass the call to Console.SetCaretPosition) causes everything to render perfectly.

Except that for some reason the Unicode Arrow char we are attempting to use doesn't render. If I use a simple char like '^' it works perfectly. 

Anyhoo, I will have a PR soon, just trying to figure out the right char to use. It's super strange it's not outputting correctly, I suspect it has something to do with my color stuff (like I have to escape it or soehitng)

**Reply by jinek** [reply]
Position of caret worked fine before. I have added caching because on some terminals it was showing big delay.

But the thing you said regarding "down" symbol can be the reason. It moves the position of caret in cache, but for some reason does not print that character, which means in reality caret skips 1 char.

**Reply by tomlm** [reply]
That makes sense. For some reason it is not displaying the char. Very weird as it works in other console app I wrote. Still poking around on it

**Reply by tomlm** [reply]
I figured it out! The Up/Down chars are unicode chars. Console app was set to ASCII.  
```csharp
            Console.OutputEncoding = System.Text.Encoding.UTF8;
```
And it all renders perfectly, on both windows and ubuntu.

**Reply by jinek** [reply]
As I remember on linux there is also some special initialization sent to TTY. I think eventualy I just need to learn how to receive debugging info from the terminal itself.
